Add support for multiple git resolver configurations

piyush-garg opened the pull request:
This will add support for providing multiple configurations
in git resolver configmap
Configurations can be provided using unique key and that key
can be passed as a param in the taskrun/pipelinerun to resolver

Old format is supported to provide backward compatibility and
docs added in details about how to use the feature.

Unit and e2e test added

Fixes https://github.com/tektoncd/pipeline/issues/5487

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Add support for multiple git resolver configurations
```

piyush-garg commented:
/test pull-tekton-pipeline-integration-tests

piyush-garg commented:
/test pull-tekton-pipeline-go-coverage-df

piyush-garg commented:
/test pull-tekton-pipeline-alpha-integration-tests

piyush-garg commented:
/test pull-tekton-pipeline-go-coverage-df

piyush-garg commented:
/test pull-tekton-pipeline-integration-tests

piyush-garg commented:
/test pull-tekton-pipeline-go-coverage-df

vdemeester reviewed (approved):
CodeQL is not happy 😅.

So I am *ok* to get this merge but it doesn't feel like something that scale that much. It will work fine for small teams, but I do feel like bigger team that would require tens or hundreds of identifier would instead use the parameter. Maybe it's just a documentation issue.

/cc @afrittoli

piyush-garg commented:
/retest

savitaashture commented on pkg/resolution/resolver/git/resolver.go:
```diff
@@ -247,9 +251,15 @@ var _ framework.TimedResolution = &Resolver{}
-func (r *Resolver) GetResolutionTimeout(ctx context.Context, defaultTimeout time.Duration) time.Duration {
-	conf := framework.GetResolverConfigFromContext(ctx)
-	if timeoutString, ok := conf[DefaultTimeoutKey]; ok {
+func (r *Resolver) GetResolutionTimeout(ctx context.Context, defaultTimeout time.Duration, params map[string]string) time.Duration {
```
I see same logic and function GetResolutionTimeout  here [pkg/remoteresolution/resolver/git/resolver.go](https://github.com/tektoncd/pipeline/pull/8263/files?show-viewed-files=true&file-filters%5B%5D=#diff-ecb3402dfe5b1846a7e8733f7018828293186ce7040d997973def09fce3d5cb7R144-R159)  as well  can we reuse instead of keeping in 2 places ?

afrittoli commented on pkg/resolution/resolver/git/config.go:
```diff
@@ -43,3 +51,54 @@ const (
+	for k, v := range conf {
+		var tokenIdentifier, tokenKey string
+		key := strings.Split(k, ".")
+		if len(key) >= 3 && key[0] == "identifier" {
```
We don't support having a "." in the name of a configuration item, we should look for `len(key)== 3`.

As mentioned before, I think the `identifier` here is redundant. This could be something like:

```golang
var configName
key := strings.Split(k, ".")
switch len(key) {
case 2:
  configName = key[0]
case 1:
  configName = "default"
default:
  // return an error
}
```

piyush-garg replied:
ack length should be 3, but i just added if someone is adding by mistake, so it is processed but it will not match any jsonTag

afrittoli replied:
I used `2` thinking ahead about my previous comment that the `identifier` part is redundant.
The non-match will be silently ignored, which is not ok I think.

piyush-garg replied:
handled this and returning error for other cases

afrittoli commented on pkg/resolution/resolver/git/config.go:
```diff
@@ -43,3 +51,54 @@ const (
+			fieldName := field.Name
+			jsonTag := field.Tag.Get("json")
+			if tokenKey == jsonTag {
+				tokenDetails := gitResolverConfig.ScmTokens[tokenIdentifier]
+				var scm interface{} = &tokenDetails
+				structValue := reflect.ValueOf(scm).Elem()
```
These three lines don't need to be within the loop over fields. You can get your `structValue` before starting the loop, update it for each field where the config name matches the json tag, and after the loop you can set it back to the `gitResolverConfig` object (so move line 99 after the loop.

Note that this implicitly let one set the same config multiple times, and the order of precedence depends on the order we get from`range conf` - we should either document this or check if a key is already set and fail.

piyush-garg replied:
this is to handle the case like while parsing, we may receive different key, so i am parsing each key one by one

afrittoli replied:
Yes, I understand that, but you have two nested loops, the outer loop checks all the config keys, each with its own `tokenIdentifier`, the inner loop checks all known config keys against the `tokenKey` extracted from `tokenIdentifier` - so within the inner loop `tokenIdentifier` remains the same, until the next line of config is processed and the inner loop is executed again with a different `tokenIdentifier`.

The following should work:

```golang
	for k, v := range conf {
		var tokenIdentifier, tokenKey string
		key := strings.Split(k, ".")
		if len(key) >= 3 && key[0] == "identifier" {
			tokenKey = strings.Join(key[2:], ".")
			tokenIdentifier = key[1]
		} else {
			tokenKey = k
			tokenIdentifier = "default"
		}
		tokenDetails, ok := gitResolverConfig.ScmTokens[tokenIdentifier]
		if !ok {
			tokenDetails = ScmInfo{}
		}
		var scm interface{} = &tokenDetails
		structValue := reflect.ValueOf(scm).Elem()
		for i := range structType.NumField() {
			field := structType.Field(i)
			jsonTag := field.Tag.Get("json")
			if tokenKey == jsonTag {
				structValue.FieldByName(field.Name).SetString(v)
			}
		}
                gitResolverConfig.ScmTokens[tokenIdentifier] = structValue.Interface().(ScmInfo)
	}
```

piyush-garg replied:
Actually, the way I have done is, it is doing all reflect thing only if the jsonTag is matching. If wrong key is provided then all those reflect code is not executed, which I think is better instead of doing it every time.

afrittoli replied:
NIT: I see, it's a small improvement in case of incorrect configuration, which should be the exception. I find the code I pasted above easier to parse, but that's probably just me 😅  No need to change it

afrittoli commented on pkg/resolution/resolver/git/config.go:
```diff
@@ -43,3 +51,54 @@ const (
+	APISecretNamespace string `json:"api-token-secret-namespace"`
+}
+
+func GetGitConfig(ctx context.Context) GitResolverConfig {
```
This function needs unit tests!

piyush-garg replied:
i tried to do this, but was not able to mock, may be will give one more try.

piyush-garg replied:
unit tests added

afrittoli commented on pkg/resolution/resolver/git/resolver.go:
```diff
@@ -247,9 +251,15 @@ var _ framework.TimedResolution = &Resolver{}
-	if timeoutString, ok := conf[DefaultTimeoutKey]; ok {
+func (r *Resolver) GetResolutionTimeout(ctx context.Context, defaultTimeout time.Duration, params map[string]string) time.Duration {
+	gitResolverConfig := GetGitConfig(ctx)
+	var conf ScmInfo
+	if tokenIdentifier, ok := params[TokenIdentifierParam]; ok {
+		conf = gitResolverConfig.ScmTokens[tokenIdentifier]
+	} else {
+		conf = gitResolverConfig.ScmTokens["default"]
+	}
```
This few lines are used 4 times in the code to extract a configuration that matches the config name provided in the parameters. In case the name is in the parameter, but it does not match the config, the `conf` object will be empty and we should error on that.

I would suggest moving this into a function that returns the error if applicable and use that function in the various places.

piyush-garg replied:
ack , will do this change

afrittoli reviewed (commented):
Thanks for this PR, it's a great functionality to add.

I left a new comments, the most important are about handling error cases and adding some code coverage to `config.go`; plus I would really like to see the redundant `identifier` go away from the configuration.

I think changing the `GitResolverConfig` from a `struct` to a `map` would simplify the code in a few places. 

I know that comments about names sound picky, so feel free to ignore if you prefer. I think it would help with code readability though as I struggled a bit to associate variables with the value they carried as I could not associated the name to the value easily.

There are a few functions that take the param map as input now, only to extract the parameter that identifies the config name, so maybe they could take a string only. On the plus side, passing the params is more future-proof, so it's probably fine to keep them as they are.

afrittoli commented:
/hold

afrittoli commented:
Sorry to hold the PR, but since the structure of the configuration and name of the parameters are user facing, we should fix them before this is merge. @vdemeester @chitrangpatel WDYT

piyush-garg commented on docs/git-resolver.md:
```diff
@@ -195,6 +191,113 @@ spec:
+  default-org: ""
+
+  # configuration 2, will be used if token-identifier params passed with value test1
+  identifier.test1.fetch-timeout: "5m"
```
i think this looks good too. Idea to add `identifier` prefix to parse only the keys with this prefix for extra configs, and also it resonates with the extra param pass with key `token-identifier` but i am fine if we want to change it like that cc @vdemeester @chitrangpatel for your thoughts.

vdemeester replied:
we could have `config.test1.…` maybe ? I understand the "need" to have a prefix so that we can easily parse them and not have possible name conflicts (if someone wants to name its configuration "scm-type" for example 😂). 

But I would also be fine to go with `<config-name>.<config-key>: value` as I don't think there will be much more conflicts.

piyush-garg replied:
Did like `<config-name>.<config-key>: value`

piyush-garg commented on docs/git-resolver.md:
```diff
@@ -195,6 +191,113 @@ spec:
+
+  # configuration 2, will be used if token-identifier params passed with value test1
+  identifier.test1.fetch-timeout: "5m"
+  identifier.test1.default-url: ""
```
yes these are the default value for params if not passed, these can be overwritten, like the case now for single config

piyush-garg commented on docs/git-resolver.md:
```diff
@@ -195,6 +191,113 @@ spec:
+      value: main
+    - name: pathInRepo
+      value: task/git-clone/0.6/git-clone.yaml
+    - name: token-identifier
```
ack regarding camelCase, i am fine to change this, lets come up with a param name. May be `configIdentifier` or `configNameKey`

afrittoli replied:
`configIdentifier` or `configKey` or `configName` all sound good to me.

piyush-garg replied:
used `configKey` now, but I remember `token-identifier` because configMap keys are like that, but yes params are camelCase. Thanks for catching this.

piyush-garg commented on pkg/resolution/resolver/git/config.go:
```diff
@@ -43,3 +51,54 @@ const (
 )
+
+type GitResolverConfig struct {
+	ScmTokens map[string]ScmInfo
```
Previously, I kept the default field straight away here and this map for additional configs. Later kept this map only, so struct is like that, but fine to change.

afrittoli replied:
Yes, please 🙏

piyush-garg replied:
did like `type GitResolverConfig map[string]ScmConfig`

piyush-garg commented on pkg/resolution/resolver/git/config.go:
```diff
@@ -43,3 +51,54 @@ const (
+	ScmTokens map[string]ScmInfo
+}
+
+type ScmInfo struct {
```
I am fine to change the change, i kept it scm because there is field `scm-type` and we provide github, gitlab etc

afrittoli replied:
I see - thanks for clarifying - fine to keep `ScmInfo` then.

piyush-garg replied:
i changed it too `type ScmConfig struct ` to have but message

piyush-garg commented on pkg/resolution/resolver/git/config.go:
```diff
@@ -43,3 +51,54 @@ const (
+
+func GetGitConfig(ctx context.Context) GitResolverConfig {
+	var scmInfo interface{} = &ScmInfo{}
+	structType := reflect.TypeOf(scmInfo).Elem()
```
i tried to do this, but was not able to mock, may be will give one more try.

afrittoli replied:
If they are only used in tests they should be moved to the test package

piyush-garg commented on pkg/resolution/resolver/git/config.go:
```diff
@@ -43,3 +51,54 @@ const (
+		var tokenIdentifier, tokenKey string
+		key := strings.Split(k, ".")
+		if len(key) >= 3 && key[0] == "identifier" {
+			tokenKey = strings.Join(key[2:], ".")
```
this is to handle the case `identifier.test.default-url` like these key in configmap

afrittoli replied:
What I meant here is that you only need re-join `key[2:]` if there are more dots than expected, e.g. if we had something like `identified.test.default.url`, which we don't and don't want to have.
If thought it would be better to fail here and say that the config is invalid.

If you let it through, as you said it won't match any of the expected config, and it will be silently ignored, which may lead users to think they have configured something while they actually didn't.

I think it's fine to keep this as it is here, as long as we fail later on for unrecognised settings.

piyush-garg replied:
did the handling for invalid key format

piyush-garg commented:
> Sorry to hold the PR, but since the structure of the configuration and name of the parameters are user facing, we should fix them before this is merge. @vdemeester @chitrangpatel WDYT

No worries regarding this. Lets all approve and agree with use facing things, then we can proceed with merge

piyush-garg commented:
> Sorry to hold the PR, but since the structure of the configuration and name of the parameters are user facing, we should fix them before this is merge. @vdemeester @chitrangpatel WDYT

Hey @afrittoli @chitrangpatel @vdemeester 

I have done all the changes based on comments, please take a look now. Thanks

afrittoli reviewed (commented):
Thank you for all the updates. 
I couple more comments, mostly cosmetic, some about error handling which I think needs to be addressed.

piyush-garg commented on docs/git-resolver.md:
```diff
@@ -195,6 +191,114 @@ spec:
 
+### Specifying Configuration for Multiple Git Providers
+
+You can specify configurations for multiple providers and even multiple configurations for same provider to use in 
```
done

piyush-garg commented on docs/git-resolver.md:
```diff
@@ -195,6 +191,114 @@ spec:
+
+You can specify configurations for multiple providers and even multiple configurations for same provider to use in 
+different tekton resources. You need to first add details in configmap with your unique identifier key prefix.
+To use them in tektonresouce, pass the unique key mentioned in configmap as an extra param to resolver with key 
```
done

piyush-garg commented on docs/git-resolver.md:
```diff
@@ -195,6 +191,114 @@ spec:
+You can specify configurations for multiple providers and even multiple configurations for same provider to use in 
+different tekton resources. You need to first add details in configmap with your unique identifier key prefix.
+To use them in tektonresouce, pass the unique key mentioned in configmap as an extra param to resolver with key 
+`configKey` and value your unique key. If no `configKey` param passed, `default` value will be considered. You can specify 
```
done

piyush-garg commented on docs/git-resolver.md:
```diff
@@ -195,6 +191,114 @@ spec:
+`configKey` and value your unique key. If no `configKey` param passed, `default` value will be considered. You can specify 
+default configuration in configmap either by mentioning no unique identifier or by using default identifier `default`
+
+### Sample Configmap
```
done

piyush-garg commented on pkg/resolution/resolver/git/config_test.go:
```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2022 The Tekton Authors
```
done

piyush-garg commented on pkg/remoteresolution/resolver/framework/reconciler.go:
```diff
@@ -118,9 +118,14 @@ func (r *Reconciler) resolve(ctx context.Context, key string, rr *v1beta1.Resolu
 	timeoutDuration := defaultMaximumResolutionDuration
 	if timed, ok := r.resolver.(framework.TimedResolution); ok {
-		timeoutDuration = timed.GetResolutionTimeout(ctx, defaultMaximumResolutionDuration)
+		timeoutDuration, _ = timed.GetResolutionTimeout(ctx, defaultMaximumResolutionDuration, paramsMap)
```
done

piyush-garg commented on pkg/remoteresolution/resolver/git/resolver.go:
```diff
@@ -141,13 +141,16 @@ var _ resolutionframework.TimedResolution = &Resolver{}
 		timeout, err := time.ParseDuration(timeoutString)
 		if err == nil {
-			return timeout
+			return timeout, nil
```
done

piyush-garg commented on pkg/resolution/resolver/framework/reconciler.go:
```diff
@@ -109,9 +109,14 @@ func (r *Reconciler) resolve(ctx context.Context, key string, rr *v1beta1.Resolu
 	timeoutDuration := defaultMaximumResolutionDuration
 	if timed, ok := r.resolver.(TimedResolution); ok {
-		timeoutDuration = timed.GetResolutionTimeout(ctx, defaultMaximumResolutionDuration)
+		timeoutDuration, _ = timed.GetResolutionTimeout(ctx, defaultMaximumResolutionDuration, paramsMap)
```
done

piyush-garg commented on pkg/resolution/resolver/git/resolver.go:
```diff
@@ -247,29 +248,36 @@ var _ framework.TimedResolution = &Resolver{}
 		timeout, err := time.ParseDuration(timeoutString)
 		if err == nil {
-			return timeout
+			return timeout, nil
```
done

piyush-garg commented:
> Thank you for all the updates. I couple more comments, mostly cosmetic, some about error handling which I think needs to be addressed.

@afrittoli I have addressed all the comments

afrittoli commented on docs/git-resolver.md:
```diff
@@ -195,6 +191,117 @@ spec:
+It is possible to specify configurations for multiple providers and even multiple configurations for same provider to use in 
+different tekton resources. You need to first add details in configmap with your unique identifier key prefix.
+To use them in tekton resources, pass the unique key mentioned in configmap as an extra param to resolver with key 
+`configKey` and value your unique key. If no `configKey` param is passed, `default` will be used. You can specify 
```
NIT: you can...

afrittoli reviewed (commented):
Thanks a lot for all the updates!
Docs could use a couple of minor updates, but that can be done as a follow-up PR, good candidate for Hacktoberfest :)
/lgtm

afrittoli commented:
/hold cancel
